sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala

```diff
@@ -40,3 +41,98 @@ object XmlExpressionEvalUtils {
     UTF8String.fromString(dataType.sql)
   }
 }
+
+object XPathEvaluatorFactory {
+  def create(dataType: DataType, path: UTF8String): XPathEvaluator = {
+    dataType match {
```

Comment:
Chasing the `dataType` doesn't look nice. How about to distribute instantiations across expressions:
- Define in `XPathExtract`:
```
  protected def evaluator: XPathEvaluator
``` 
- and override it childs:
```scala
  @transient override lazy val evaluator = XPathBooleanEvaluator(pathUTF8String)
```